replication/row_event.go

```diff
@@ -195,9 +294,159 @@ func (e *TableMapEvent) Dump(w io.Writer) {
+	fmt.Fprintf(w, "CollationMap: %#v\n", collationMap)
+
+	nameMaxLen := 0
+	for _, name := range e.ColumnName {
+		if len(name) > nameMaxLen {
+			nameMaxLen = len(name)
+		}
+	}
+	nameFmt := "  %s"
+	if nameMaxLen > 0 {
+		nameFmt = fmt.Sprintf("  %%-%ds", nameMaxLen)
+	}
+
+	primaryKey := map[int]struct{}{}
+	for _, pk := range e.PrimaryKey {
+		primaryKey[int(pk)] = struct{}{}
+	}
+
+	fmt.Fprintf(w, "Columns: \n")
+	for i := 0; i < int(e.ColumnCount); i++ {
+		if len(e.ColumnName) == 0 {
+			fmt.Fprintf(w, nameFmt, "<n/a>")
+		} else {
+			fmt.Fprintf(w, nameFmt, e.ColumnName[i])
+		}
+
+		fmt.Fprintf(w, "  type=%-3d", e.ColumnType[i])
+
+		if IsNumericType(e.ColumnType[i]) {
+			if len(unsignedMap) == 0 {
+				fmt.Fprintf(w, "  unsigned=<n/a>")
+			} else if unsignedMap[i] {
+				fmt.Fprintf(w, "  unsigned=yes")
+			} else {
+				fmt.Fprintf(w, "  unsigned=no ")
+			}
+		} else if IsCharacterType(e.ColumnType[i]) {
+			if len(collationMap) == 0 {
+				fmt.Fprintf(w, "  collation=<n/a>")
+			} else {
+				fmt.Fprintf(w, "  collation=%d ", collationMap[i])
+			}
+		}
+
+		available, nullable := e.Nullable(i)
+		if !available {
+			fmt.Fprintf(w, "  null=<n/a>")
+		} else if nullable {
+			fmt.Fprintf(w, "  null=yes")
+		} else {
+			fmt.Fprintf(w, "  null=no ")
+		}
+
+		if _, ok := primaryKey[i]; ok {
+			fmt.Fprintf(w, "  pri")
+		}
+
+		fmt.Fprintf(w, "\n")
+	}
 	fmt.Fprintln(w)
 }
 
+// Nullable returns the nullablity of the i-th column.
+// If null bits are not available, available is false.
+// i must be in range [0, ColumnCount).
+func (e *TableMapEvent) Nullable(i int) (available, nullable bool) {
+	if len(e.NullBitmap) == 0 {
+		return
+	}
+	return true, e.NullBitmap[i/8]&(1<<uint(i%8)) != 0
+}
+
+// UnsignedMap returns a map: column index -> unsigned.
+// Note that only numeric columns will be returned.
+// nil is returned if not available or no numeric columns at all.
+func (e *TableMapEvent) UnsignedMap() map[int]bool {
+	if len(e.SignednessBitmap) == 0 {
+		return nil
+	}
+	p := 0
+	ret := make(map[int]bool)
+	for i := 0; i < int(e.ColumnCount); i++ {
+		if !IsNumericType(e.ColumnType[i]) {
+			continue
+		}
+
+		ret[i] = e.SignednessBitmap[p/8]&(1<<uint(7-p%8)) != 0
+		p++
+	}
+	return ret
+}
+
+// CollationMap returns a map: column index -> collation id.
+// Note that only character columns will be returned.
+// nil is returned if not available or no character columns at all.
+func (e *TableMapEvent) CollationMap() map[int]uint64 {
+
+	if len(e.DefaultCharset) != 0 {
+		defaultCollation := e.DefaultCharset[0]
+
+		// character column index -> collation
+		collations := make(map[int]uint64)
+		for i := 1; i < len(e.DefaultCharset); i += 2 {
+			collations[int(e.DefaultCharset[i])] = e.DefaultCharset[i+1]
+		}
+
+		p := 0
+		ret := make(map[int]uint64)
+		for i := 0; i < int(e.ColumnCount); i++ {
+			if !IsCharacterType(e.ColumnType[i]) {
+				continue
+			}
+
+			if collation, ok := collations[p]; ok {
```

Comment:
Can we use `collations[i]` here?

Reply:
hmm... `i` and `p` are two different counters in this loop: `i` is the index of all columns while `p` is the index of character columns, here is an example to illustrate:

| column | i | p |
|---|---|---|
| a int | 0 | n/a |
| b char(8) | 1 | 0 |
| c bigint | 2 | n/a |
| d char(8) | 3 | 1 |

Reply:
👍 you're right, I found `The column index is counted only in all character columns.` now.